extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/DefaultTenantConfigResolver.java

```diff
@@ -113,33 +114,37 @@ public OidcTenantConfig apply(OidcTenantConfig tenantConfig) {
                 });
     }
 
+    Uni<TenantConfigContext> resolveContext(String tenantId) {
```

Comment:
It highlights an interesting problem. If the token is arriving out of band, bypassing the OIDC authentication mechanism, it just won't work because the tenant resolution nearly always needs RoutingContext.

So my proposal is as follows - `resolveContext(String tenantId)` should check the static or dynamic map or default tenant and if the tenant config context is not there - report the exception, the resolution must have happened already by the time the call has been made. 
Then we should doc that if the users want to use this feature in a multi-tenant setup, they must inject `DefaultTenantConfigResolver` alongside RoutingContext and call its `resolveContext(routingContext)` while the routing context is still available.

This also highlights another thing, if there is only default tenant configured, the feature should work without having to apply `@TenantFeature` - and in this case only a default tenant should be checked

Reply:
Perhaps a Draft ? It is close to being merged but we might need to tune a few things before that

Reply:
Though may be with static tenants it can also work as the are all mapped at startup, yeah. But for dynamic tenants it won't work so lets have a message what to do when it is a dynamic tenant resolution but throw the exception if the tenant context is not available

Reply:
@sberyozkin I realize some of comments below want make sense to you, but I need to push back in order to understand.

> Perhaps a Draft ? It is close to being merged but we might need to tune a few things before that

If there are concrete scenarios that don't work or need to be improved, I need to create tests for them first. I need to understand requirements. Putting something to draft is not going to solve this. If current test coverage or scope of the PR didn't comply the opened issue, draft would make sense.

That said, please take your time to think about it, as far as I'm concerned take month, I'm fine!

> This also highlights another thing, if there is only default tenant configured, the feature should work without having to apply @TenantFeature - and in this case only a default tenant should be checked

You can only produce beans at build time, while tenants are configured at run time. How would you do that? And even default tenant can have `quarkus.oidc.tenant-id` and if it is not set, you can inject default tenant using default name as I do. We need to know what tenant is user trying to inject, this can't be done at runtime unless you want `TenantIdentityProvider` to be singleton for all tenants and you will need to specify tenant id everytime you call `authenticate` (or some other version of this).

> So my proposal is as follows - resolveContext(String tenantId) should check the static or dynamic map or default tenant and if the tenant config context is not there - report the exception, 

This works for static tenant the same way as original mechanism - it is extracted into methods and is sharing the same code. When tenant config context is not there, default tenant config context is used. Why should it work differently? If tenant id is selected during `RoutingContext` processing and the context is not there, default one is used.

> the resolution must have happened already by the time the call has been made.

I think should work the same way as `OidcIdentityProvider` does - only resolve tenant config on demand, because only this way you know that when context is created by `TenantConfigContextFactory` everything is supposed to be ready?

> Though may be with static tenants it can also work as the are all mapped at startup, yeah.

Bit lost, I have tests right, what do you mean it may work :-) Maybe I'm not testing it properly? If so, please shout!

> But for dynamic tenants it won't work so lets have a message what to do when it is a dynamic tenant resolution but throw the exception if the tenant context is not available

I don't have a means to find out that selected tenant has dynamic tenant resolution. It can't work based on `RoutingContext` because this feature is all about working without `RoutingContext`, so isn't that something that user already knows?

Reply:
@michalvavrik You are welcome to push back as much as you like, this is a nature of the healthy review process :+1:

Well, I'm sorry the original issue did not lay out the precise requirements, so this is why we tune things now during this review, you've given it the best shot, but I'm seeing some things that IMHO should be changed, so related to this specific tenant resolution changes, here is a summary of what I'm thinking about:

1. If I only have, say for a bearer token case,
```
quarkus.oidc.auth-server-url=someurl
```
then having to write `@Inject @TenantFeature("Default") TenantIdentityProvider provider;` looks strange to me, it is likely to be the main case, so my proposal is that when no  `@TenantFeature("Default")` qualifier is available then, internally, only the default tenant is picked up.

2. 

>I don't have a means to find out that selected tenant has dynamic tenant resolution, there is no such list, it depends on RoutingContext. It can't work based on RoutingContext because this feature is all about working without RoutingContext, so isn't that something that user already knows?

Which is why I'm saying if users want to use this feature with the dynamic tenant resolution, we should advise them in the docs that they must run the resolution while the routing context is available - that will populate the dynamic map, which in turn make it available to TenantIdentityProvider at a later stage. The only problem is that `DefaultTenantConfigResolver` is not really an API we want users to deal with directly, so may be it is worth introdcing `TenantConfigResolverManager` with a single `resolve(RoutingContext)` method - I can deal with it myself later, for now though we just need to mention IMHO that this feature does not work when a dynamic multi-tenant resolution is enabled

Reply:
> 1. ...

Fine.

> Which is why I'm saying if users want to use this feature with the dynamic tenant resolution, we should advise them in the docs that they must run the resolution while the routing context is available

Okay, I definitely didn't get this before. Fine so far.

> that will populate the dynamic map, which in turn make it available to TenantIdentityProvider at a later stage. The only problem is that DefaultTenantConfigResolver is not really an API we want users to deal with directly, so may be it is worth introducing TenantConfigResolverManager with a single resolve(RoutingContext) method - I can deal with it myself later, for now though we just need to mention IMHO that this feature does not work when a dynamic tenant resolution is used

This seems clumsy. I can easily populate auth request with `RoutingContext` when it is available and request scope is activated. 

> TenantConfigResolverManager

IMO it is not easy for users to keep track with all the classes and features. I know you know them :-) and they can look it up, but you can say that about literally every feature and framework. Let's add `RoutingContext` when available ourselves.

> for now though we just need to mention IMHO that this feature does not work when a dynamic tenant resolution is used

Fine if my suggestion above is no go.

Reply:
>I can easily populate auth request with RoutingContext when it is available and request scope is activated.

Not sure about it, the request has already gone, typically, by the time this feature is used. Am I missing something ?

Reply:
> > I can easily populate auth request with RoutingContext when it is available and request scope is activated.
> 
> Not sure about it, the request has already gone, typically, by the time this feature is used. Am I missing something ?

- Normally when auth happens `RoutingContext` can't yet be accessed from CDI, however we can try if and only if is CDI request scope active and `RoutingContext` is resolvable, we can simply put it there. Where else do you expect users to acquire `RoutingContext`?
- Let's imagine users acquire `RoutingContext` somehow else, it is also stored on duplicated context, right? In this case they can pass it like this
```
Uni<SecurityIdentity> authenticate(AccessTokenCredential token);
Uni<SecurityIdentity> authenticate(AccessTokenCredential token, RoutingContext context);
```

I'm fine with your way, please let me know ASAP @sberyozkin  (sorry!)

Reply:
How are users going to get `RoutingContext` when HTTP request has already ended and this happens on different CDI request context. Only idea I have is that you expect users to store it and use it later?

Reply:
> Which is why I'm saying if users want to use this feature with the dynamic tenant resolution, we should advise them in the docs that they must run the resolution while the routing context is available

Ah, I think I understand. You want this

```
 // check if the connection has already been created
            TenantConfigContext readyTenantContext = tenantConfigBean.getDynamicTenantsConfig()
                    .get(tenantContext.oidcConfig.tenantId.get());
```
to work, right?

Reply:
@michalvavrik Sorry for the confusion if any but I did not suggest to bring `RoutingContext`,

>Only idea I have is that you expect users to store it and use it later?

No, I proposed, for now, mention in the docs that this feature is not supported with dynamic tenants, and then later, tell uses who want to use the dynamic resolution alongside this feature, is that they only use RoutingContext to resolve the tenant context - which would have a side-effect of a dynamic tenant map with some tenant id populated.

I think your idea of having an overloaded method accepting RoutingContext is good - because that will let users run bearer token verification on tokens coming out of band, when this context is visible - but somehow it feels may be we should avoid bringing RoutingContext into this feature at all.

May be we should follow up with the updates related the dynamic resolution later, and avoid trying to cover it now, it will probably be needed for a 1% of all cases as far as the use of this feature is concerned.

Does it sounds good to you ? The only question, if tenant configuration is not available, then `TenantIdentityProvider` should throw the exception ?

Reply:
All fine, I believe I understand now and it makes sense.

> Does it sounds good to you ? The only question, if tenant configuration is not available, then TenantIdentityProvider should throw the exception ?

This is kinda your area, because AFAICT from the code I read, normally default tenant context is used even when tenant is dynamically selected by `TokenCustomizer` (by providing tenant config with id). Is this somewhere documented or what is expected behavior?

Reply:
I can throw exception, technically it is not an issue.

Reply:
I believe I am done. I also believe I can follow-up with PR for dynamic context once you are sure `TenantConfigResolverManager`. This is good opportunity to get hands dirty with OIDC even for someone like me.

Reply:
@michalvavrik Good stuff, let me resolve this long thread and return with new very minor suggestions :-)

Reply:
Go for it :-D